installer/installer.go

```diff
@@ -0,0 +1,743 @@
+	Creds      aws.CredentialsProvider `json:"-"`
+	StackID    string                  `json:"stack_id,omitempty"`
+	StackName  string                  `json:"stack_name,omitempty"`
+	Stack      *cloudformation.Stack   `json:"-"`
+	SSHKey     *sshkeygen.SSHKey       `json:"-"`
+	SSHKeyName string                  `json:"ssh_key_name,omitempty"`
+
+	DiscoveryToken string   `json:"discovery_token"`
+	InstanceIPs    []string `json:"instance_ips,omitempty"`
+	DNSZoneID      string   `json:"dns_zone_id,omitempty"`
+
+	ControllerKey       string `json:"controller_key,omitempty"`
+	DashboardLoginToken string `json:"dashboard_login_token,omitempty"`
+
+	EventChan chan *Event   `json:"-"`
+	ErrChan   chan error    `json:"-"`
+	Done      chan struct{} `json:"-"`
+
+	eventMutex   sync.Mutex
+	errMutex     sync.Mutex
+	persistMutex sync.Mutex
+
+	cf        *cloudformation.CloudFormation
+	ec2Client *ec2.EC2
+
+	instances int
+}
+
+func RunAWS(region string, instances int, creds aws.CredentialsProvider) *Stack {
+	s := &Stack{
+		EventChan: make(chan *Event),
+		ErrChan:   make(chan error),
+		Done:      make(chan struct{}),
+
+		InstanceIPs: make([]string, 0, instances),
+		instances:   instances,
+
+		Creds: creds,
+
+		ec2Client: ec2.New(creds, region, nil),
+		cf:        cloudformation.New(creds, region, nil),
+	}
+	s.load()
+	if s.ImageID != "" && region != s.Region {
+		s.ImageID = ""
+	}
+	s.Region = region
+
+	go func() {
+		defer close(s.Done)
+
+		// TODO(jvatic): keep track of what step we're on (and related data) in `s` so we can avoid re-creating usable components
+
+		steps := []func() error{
+			s.fetchImageID,
+			s.allocateDomain,
+			s.createKeyPair,
+			s.createStack,
+			s.fetchStackOutputs,
+			s.configureDNS,
+			s.healthCheckInstances,
+			s.bootstrap,
+		}
+
+		for _, step := range steps {
+			if err := step(); err != nil {
+				s.persist()
+				s.HandleError(err)
+				return
+			}
+			if err := s.persist(); err != nil {
+				s.HandleError(err)
+			}
+		}
+	}()
+
+	return s
+}
+
+func (s *Stack) HandleEvent(name, description string) {
+	s.eventMutex.Lock()
+	defer s.eventMutex.Unlock()
+	s.EventChan <- &Event{name, description}
+}
+
+func (s *Stack) HandleError(err error) {
+	s.errMutex.Lock()
+	defer s.errMutex.Unlock()
+	s.ErrChan <- err
+}
+
+func (s *Stack) fetchImageID() error {
+	go s.HandleEvent("image_id_fetch", "Fetching image manifest")
+	if err := func() error {
+		latestVersion, err := fetchLatestVersion()
+		if err != nil {
+			return err
+		}
+		var imageID string
+		for _, i := range latestVersion.Images {
+			if i.Region == s.Region {
+				imageID = i.ID
+				break
+			}
+		}
+		if imageID == "" {
+			err := errors.New(fmt.Sprintf("No image found for region %s", s.Region))
+			return err
+		}
+		s.ImageID = imageID
+		return nil
+	}(); err != nil {
+		go s.HandleEvent("image_id_fetch_error", err.Error())
+		if s.ImageID != "" {
+			go s.HandleEvent("image_id_load", "Falling back to saved Image ID")
```

Comment:
It looks like these should be channel sends, otherwise the ordering is not guaranteed.